src/main/java/org/scijava/plugin/AbstractSingletonService.java

```diff
@@ -84,20 +85,63 @@ public ObjectService objectService() {
 		return (P) instanceMap.get(pluginClass);
 	}
 
+//-- Event handlers --
+
+	@EventHandler
+	protected void onEvent(final PluginsRemovedEvent event) {
+		if (instanceMap == null) return;
+		for (final PluginInfo<?> info : event.getItems()) {
+			final PT obj = instanceMap.remove(info.getPluginClass());
+			if (obj != null) { // we actually removed a plugin
+				instances.remove(obj);
+				objectService.removeObject(obj);
+			}
+		}
+	}
+
+	@EventHandler
+	protected void onEvent(final PluginsAddedEvent event) {
+		if (instanceMap == null) return;
+		// collect singleton plugins
+		final List<PluginInfo<PT>> singletons = new ArrayList<>();
+		for (final PluginInfo<?> pluginInfo : event.getItems()) {
+			if (SingletonPlugin.class.isAssignableFrom(pluginInfo.getPluginType())) {
```

Comment:
I believe this should be `getPluginType().isAssignableFrom(pluginInfo.getPluginType())`.

Could you please fix it, and then update the unit test to prove that when we add plugins of multiple types, only the correct ones are picked up by the respective services?

Reply:
I added a test for this and it was indeed a problem, so I fixed it as you suggested.